core/src/main/java/com/netflix/iceberg/LocationProviders.java

```diff
@@ -0,0 +1,114 @@
+
+public class LocationProviders {
+
+  public static LocationProvider locationsFor(String location, Map<String, String> properties) {
+    if (PropertyUtil.propertyAsBoolean(properties,
+        TableProperties.OBJECT_STORE_ENABLED,
+        TableProperties.OBJECT_STORE_ENABLED_DEFAULT)) {
+      return new ObjectStoreLocationProvider(location, properties);
+    } else {
+      return new DefaultLocationProvider(location, properties);
+    }
+  }
+
+  static class DefaultLocationProvider implements LocationProvider {
+    private final String dataLocation;
+
+    DefaultLocationProvider(String tableLocation, Map<String, String> properties) {
+      this.dataLocation = stripTrailingSlash(properties.getOrDefault(
+          TableProperties.WRITE_NEW_DATA_LOCATION,
+          String.format("%s/data", tableLocation)));
+    }
+
+    @Override
+    public String newDataLocation(PartitionSpec spec, StructLike partitionData, String filename) {
+      return String.format("%s/%s/%s", dataLocation, spec.partitionToPath(partitionData), filename);
+    }
+
+    @Override
+    public String newDataLocation(String filename) {
+      return String.format("%s/%s", dataLocation, filename);
+    }
+  }
+
+  static class ObjectStoreLocationProvider implements LocationProvider {
+    private static final Transform<String, Integer> HASH_FUNC = Transforms
+        .bucket(Types.StringType.get(), Integer.MAX_VALUE);
+
+    private final String storageLocation;
+    private final String context;
+
+    ObjectStoreLocationProvider(String tableLocation, Map<String, String> properties) {
+      this.storageLocation = stripTrailingSlash(properties.get(OBJECT_STORE_PATH));
+      this.context = pathContext(tableLocation);
+    }
+
+    @Override
+    public String newDataLocation(PartitionSpec spec, StructLike partitionData, String filename) {
+      return newDataLocation(String.format("%s/%s", spec.partitionToPath(partitionData), filename));
+    }
+
+    @Override
+    public String newDataLocation(String filename) {
+      int hash = HASH_FUNC.apply(filename);
+      return String.format("%s/%08x/%s/%s", storageLocation, hash, context, filename);
+    }
+
+    private static String pathContext(String tableLocation) {
+      Path dataPath = new Path(tableLocation);
+      Path parent = dataPath.getParent();
+      String resolvedContext;
+      if (parent != null) {
+        // remove the data folder
+        resolvedContext = String.format("%s/%s", parent.getName(), dataPath.getName());
+      } else {
+        resolvedContext = dataPath.getName();
+      }
+
+      Preconditions.checkState(
+          !resolvedContext.endsWith("/"),
+          "Path context must not end with a slash.");
+
+      return resolvedContext;
+    }
+  }
+
+  private static String stripTrailingSlash(String path) {
+    String result = path;
+    while (result.endsWith("/")) {
+      result = result.substring(0, path.length() - 1);
```

Comment:
I think this is a bug - this may not work as expected, I think it goes into an infinite loop for strings ending with more than one slash, i.e. `stripTrailingSlash("trailing/slash//////")` cause path.length doesn't change when iterating right?

Rewriting this line as `result = result.substring(0, result.length() - 1);` we'll have the method converge to the expected output as 

```code
trim result=trailing/slash//////
trim result=trailing/slash/////
trim result=trailing/slash////
trim result=trailing/slash///
trim result=trailing/slash//
trim result=trailing/slash/
trailing/slash
```

Reply:
I agree. I'll fix this in the PR. Thanks for reviewing!